internal/datastore/mongo_store.go

```
@@ -2,89 +2,131 @@ package datastore
-        uri = val
+	// Use local development mongodb instance if env variable not set
+	uri := "mongodb://root:example@localhost:27017/?maxPoolSize=20&w=majority"
+
+	if val, ok := os.LookupEnv("MONGO_URI"); ok {
+		uri = val
 	}
 
-    log.Println("Connecting to MongoDB...")
+	log.Println("Connecting to MongoDB...")
+
+	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
 
-    client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
-    
-    if err != nil {
-        log.Fatal(err)
-    }
+	if err != nil {
+		log.Fatal(err)
+	}
 
-    new_mongodb := MongoDB {
-        client,
-        client.Database("simplq").Collection("queue"),
-        client.Database("simplq").Collection("token"),
-    }
+	new_mongodb := MongoDB{
+		client,
+		client.Database("simplq").Collection("queue"),
+		client.Database("simplq").Collection("token"),
+	}
 
-    log.Println("Successfully connected to MongoDB!")
+	log.Println("Successfully connected to MongoDB!")
 
-    return &new_mongodb
+	return &new_mongodb
 }
 
 func (mongodb MongoDB) CreateQueue(queue models.Queue) models.QueueId {
-    // Set id to empty so its generated by mongoDB
-    queue.Id = ""
+	// Set id to empty so its generated by mongoDB
+	queue.Id = ""
 
-    result, err := mongodb.Queue.InsertOne(context.TODO(), queue)
+	result, err := mongodb.Queue.InsertOne(context.TODO(), queue)
 
-    if err != nil {
-        log.Fatal(err)
-    }
- 
-    stringId := result.InsertedID.(primitive.ObjectID).Hex()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	stringId := result.InsertedID.(primitive.ObjectID).Hex()
 
-    return models.QueueId(stringId)
+	return models.QueueId(stringId)
 }
 
-func (mongodb MongoDB) ReadQueue(models.QueueId) models.Queue {
-    panic("Not implemented")
+func (mongodb MongoDB) ReadQueue(id models.QueueId) models.Queue {
+	result := models.Queue{}
+
+	// convert id string to ObjectId
+	queueId, _ := primitive.ObjectIDFromHex(string(id))
```

Should the error return object be caught here and checked as well for invalid object ids? Have you omitted the check because it will be caught by the next error check? Maybe it's slightly better to catch it and throw an error for invalid id string passed to avoid DB queries when not needed.

---

Checks are being done in the calling method. We can discuss further if that needs improvements/amends.

---

Got it, that works.